File: backend/src/main/java/com/example/vehicleservice/domain/model/ServiceRequest.java
package com.example.vehicleservice.domain.model;

import com.example.vehicleservice.domain.exception.DomainRuleViolationException;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;

import java.time.Instant;
import java.util.UUID;

@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class ServiceRequest {
    private UUID id;
    private UUID vehicleId;
    private String description;
    private Priority priority;
    private ServiceRequestStatus status;
    private String assignedTechnician;
    private Instant createdAt;
    private Instant completedAt;

    public static ServiceRequest create(
        UUID vehicleId,
        String description,
        Priority priority) {

        if (ValidationUtils.isBlank(vehicleId))
            throw new IllegalArgumentException("vehicleId is required");
        if (ValidationUtils.isBlank(description))
            throw new IllegalArgumentException("description is required");
        if (ValidationUtils.isBlank(priority))
            throw new IllegalArgumentException("priority is required");

        return new ServiceRequest(UUID.randomUUID(), vehicleId, description, priority, ServiceRequestStatus.OPEN, null, Instant.now(), null);
    }

    public static ServiceRequest restore(UUID id, UUID vehicleId, String description, Priority priority, ServiceRequestStatus status, String assignedTechnician, Instant createdAt, Instant completedAt) {
        return new ServiceRequest(id, vehicleId, description, priority, status, assignedTechnician, createdAt, completedAt);
    }

    public boolean isActive() {
        return this.status == ServiceRequestStatus.OPEN || this.status == ServiceRequestStatus.IN_PROGRESS || this.status == ServiceRequestStatus.WAITING_FOR_PARTS;
    }

    public void assignTechnician(String technician) {
        if(!this.isActive()) throw new DomainRuleViolationException("Only active service requests can be assigned");
        this.assignedTechnician = technician;
    }

    public void updateStatus(ServiceRequestStatus status) {
        this.status = status;
    }
}

File: .pnpm-store/v11/projects/c81b21372cff0f59f94bca07e4c61e68/src/types/serviceRequest.ts
export type Priority = 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT';

export type ServiceRequestStatus =
  | 'OPEN'
  | 'IN_PROGRESS'
  | 'WAITING_FOR_PARTS'
  | 'COMPLETED'
  | 'CANCELLED';

export interface ServiceRequest {
  id?: string;
  vehicleId: string;
  description: string;
  priority: Priority;
  status?: ServiceRequestStatus;
  assignedTechnician?: string;
  createdAt?: string;
  completedAt?: string;
}

export interface OpenServiceRequestCommand {
  vehicleId: string;
  description: string;
  priority: Priority;
}

export interface ServiceRequestFilters {
  status?: ServiceRequestStatus;
  priority?: Priority;
  regNum?: string;
  page?: number;
  size?: number;
}
